src/liboslexec/llvm_gen.cpp

```diff
@@ -160,16 +161,6 @@ BackendLLVM::llvm_run_connected_layers(Symbol& sym, int symindex, int opnum,
         const Connection& con(inst()->connection(c));
         // If the connection gives a value to this param
         if (con.dst.param == symindex) {
-            // already_run is a set of layers run for this particular op.
-            // Just so we don't stupidly do several consecutive checks on
-            // whether we ran this same layer. It's JUST for this op.
-            if (already_run) {
-                if (already_run->count(con.srclayer))
-                    continue;  // already ran that one on this op
-                else
-                    already_run->insert(con.srclayer);  // mark it
-            }
-
```

Comment:
Can you explain a bit on what the difference between the removed code block with the "local" already_run set and the m_layers_already run?  And how removing the block improves things?

Reply:
Definitely!

The "local" `already_run` is restricted in scope to preventing redundant checks originating from a single useparam op. The new `m_call_layers_inserted` set will prevent all redundant checks in the scope of an entire basic block.

Consider this shader:
```
shader example_shader(ConnectionData my_input = {})
{
    if (u > 0.0) {
        float r = my_input.x * my_input.y;
        Ci = color(r, 0, 0) * emission();
    } else {
        float r = my_input.x * u;
        float g = my_input.y * v;
        Ci = color(r, g, 0) * emission();
    }
}
```

Its control flow graph looks like this:

![plot-cfg](https://user-images.githubusercontent.com/23771/230478642-9c3e4bdf-d6e0-4b82-86ee-d925fc5d4fef.png)

The existing `already_run` check would catch the "then" redundancy (block-3), but not the "else" redundancy (block-4). `m_call_layers_inserted` catches both- hence the removal of `already_run`.

`m_layers_already_run` is scoped to the entire main section of the instance, but only catches cases outside of conditionals. That's why all my examples are wrapped in  if-statements.

Reply:
This may be the most interesting reply to a review comment in this project's history. The visualization of the oso IR is extremely helpful.

Is the script or whatever you used to generate that graph something that makes sense to check into the project? I can imagine it being really helpful for a variety of debugging tasks, not to mention documenting internals of the shader optimization.